plugins/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/IngestGeoIpPlugin.java

```diff
@@ -119,4 +119,75 @@ public void close() throws IOException {
         }
     }
 
+    /**
+     * The in-memory cache for the geoip data. There should only be 1 instance of this class..
+     * This cache differs from the maxmind's {@link NodeCache} such that this cache stores the deserialized Json objects to avoid the
+     * cost of deserialization for each lookup (cached or not). This comes at slight expense of higher memory usage, but significant
+     * reduction of CPU usage.
+     */
+    class GeoIpCache {
+        private final Cache<CacheKey, AbstractResponse> cache;
+
+        //package private for testing
+        GeoIpCache(long maxSize) {
+            if (maxSize < 0) {
+                throw new IllegalArgumentException("geoip cache size must be 0 or greater");
+            }
+            this.cache = CacheBuilder.<CacheKey, AbstractResponse>builder().setMaximumWeight(maxSize).build();
+        }
+
+        <T extends AbstractResponse> T putIfAbsent(InetAddress ip, Class<T> responseType,
+                                                   Function<InetAddress, AbstractResponse> retrieveFunction) {
+
+            //can't use cache.computeIfAbsent due to the elevated permissions for the jackson (run via the cache loader)
+            CacheKey<T> cacheKey = new CacheKey<>(ip, responseType);
+            //intentionally non-locking for simplicity...it's OK if we re-put the same key/value in the cache during a race condition.
+            AbstractResponse response = cache.get(cacheKey);
+            if (response == null) {
+                response = retrieveFunction.apply(ip);
+                cache.put(cacheKey, response);
+            }
+            return responseType.cast(response);
+        }
+
+        //only useful for testing
+        <T extends AbstractResponse> T get(InetAddress ip, Class<T> responseType) {
+            CacheKey<T> cacheKey = new CacheKey<>(ip, responseType);
+            return responseType.cast(cache.get(cacheKey));
+        }
+
+         /**
+         * The key to use for the cache. Since this cache can span multiple geoip processors that all use different databases, the response
+         * type is needed to be included in the cache key. For example, if we only used the IP address as the key the City and ASN the same
+         * IP may be in both with different values and we need to cache both. The response type scopes the IP to the correct database
+         * provides a means to safely cast the return objects.
+         * @param <T> The AbstractResponse type used to scope the key and cast the result.
+         */
+        private class CacheKey<T extends AbstractResponse> {
```

Comment:
can't this be static?

Reply:
I don't think it makes sense to allow instances of the this class to exist outside the IngestGeoIpPlugin.GeoIpCache class which it is nested. Hence not static.